supporter-product-data/src/main/scala/com/gu/lambdas/UpdateDynamoLambda.scala

```diff
@@ -0,0 +1,113 @@
+import scala.concurrent.{Await, Future}
+
+trait TimeOutCheck {
+  def timeRemainingMillis: Int
+}
+
+class ContextTimeOutCheck(context: Context) extends TimeOutCheck {
+  override def timeRemainingMillis = context.getRemainingTimeInMillis
+}
+
+class UpdateDynamoLambda extends Handler[UpdateDynamoState, UpdateDynamoState] {
+  override protected def handlerFuture(input: UpdateDynamoState, context: Context) = {
+    SafeLogger.info(s"Starting write to dynamo task for ${input.recordCount} records from ${input.filename}")
+    writeToDynamo(Stage.fromEnvironment, input, new ContextTimeOutCheck(context))
+  }
+}
+
+object UpdateDynamoLambda {
+  val batchSize = 10
+  val timeoutBufferInMillis = batchSize * 5 * 1000
+
+  def writeToDynamo(stage: Stage, state: UpdateDynamoState, timeOutCheck: TimeOutCheck): Future[UpdateDynamoState] = {
+    val csvStream = S3Service.streamFromS3(stage, state.filename)
+    val csvReader = csvStream.asCsvReader[SupporterRatePlanItem](rfc.withHeader)
+    val dynamoDBService = DynamoDBService(stage)
+
+    val unProcessed = getUnprocessedItems(csvReader, state.processedCount)
+
+    val validUnprocessed = unProcessed.filter { case (itemOrError, _) => itemOrError.isRight }
+    val invalidUnprocessed = unProcessed.filter { case (itemOrError, _) => itemOrError.isLeft }
+
+    if (invalidUnprocessed.nonEmpty)
+      SafeLogger.error(
+        scrub"There were ${invalidUnprocessed.length} CSV read failures from file ${state.filename} with line numbers ${invalidUnprocessed.map(_._2).mkString(",")}"
+      )
+
+    val batches = validUnprocessed
+      .map { case (itemOrError, index) => (itemOrError.right.get, index) }
+      .grouped(batchSize)
+
+    val processedCount = writeBatchesUntilTimeout(
+      state.processedCount,
+      batches,
+      timeOutCheck,
+      dynamoDBService
+    )
+
+    Future.successful(state.copy(
+        processedCount = processedCount
+      ))
+  }
+
+  def getUnprocessedItems(csvReader:  CsvReader[ReadResult[SupporterRatePlanItem]], processedCount: Int) =
+    csvReader.zipWithIndex.drop(processedCount).toList
+
+  def writeBatchesUntilTimeout(
+    processedCount: Int,
+    groups: Iterator[List[(SupporterRatePlanItem, Int)]],
+    timeOutCheck: TimeOutCheck,
+    dynamoDBService: DynamoDBService
+  ): Int =
+    groups.foldLeft(processedCount) {
+      (processedCount, group) =>
+        if (timeOutCheck.timeRemainingMillis < timeoutBufferInMillis) {
+          SafeLogger.info(
+            s"Aborting processing - time remaining: ${timeOutCheck.timeRemainingMillis / 1000} seconds, buffer: ${timeoutBufferInMillis / 1000} seconds"
+          )
+          return processedCount
+        }
+        SafeLogger.info(
+          s"Continuing processing - time remaining: ${timeOutCheck.timeRemainingMillis / 1000} seconds, buffer: ${timeoutBufferInMillis / 1000} seconds"
+        )
+
+        Await.result(writeBatch(group, dynamoDBService), 120.seconds)
```

Comment:
foldLeft on an iterator cant terminate early without the return trick which I guess is why you needed the Await?  awaiting will block an extra thread and also will throw any failed futures

I can't help thinking that we're trying to reimplement error recovery and replay ourselves.  What would the cloud native version look like?  Run the query and load the CSV into a sqs queue or kinesis stream, then process the stream/queue until it is empty?  Or would we just run the query from scratch and reprocess everything if something goes wrong?

Reply:
The issue is the lambda timing out when we have a large number of records, that would be an problem whether we write the items to a queue or directly to dynamo. The only other way round it would be to use something other than a lambda/step function like aws glue but there are other disadvantages to that such as the modularity of the codebase and familiarity of the tech to people who will need to maintain this code. I think this is approach gives us a decent compromise.